Issue/14129 topics you might like

zwarm opened the pull request:
Fixes #14129

This PR redesigns the "Topics you like" section" of the Discover feed by swapping out the vertical topic text list with a horizontal topic chip list.

Highlights include
- `ReaderInterestsCardViewHolder`
 -- Updated the recyclerView from Vertical to Horizontal
 -- Attached an `OnItemTouchListener` to the recyclerView to handle the chip scrolling because the topics Horizontal recyclerView is contained within a Vertical recyclerView which is then contained in a ViewPager2. We need to have the inner most recycler view scroll correctly, while still allowing tab scrolling. 
- `ReaderPostUiStateBuilder`
-- Logic was added to build the color style profile for the chips
-- There are currently 4 fixed colors and if the list contains more than 4, the colors will recycle. Note: The max topics allowed at the time of this PR is 5
 
|light |dark|scrolling
|---|---|--|
|![light](https://user-images.githubusercontent.com/506707/109540243-3354b200-7a90-11eb-9a3a-d9ea83825494.png)|![dark](https://user-images.githubusercontent.com/506707/109540238-318aee80-7a90-11eb-9c29-fd02dd69afa2.png)|![scroll](https://user-images.githubusercontent.com/506707/109540774-ef15e180-7a90-11eb-8c36-1651782cbddc.gif)

*To Test*
- Launch the app
- Navigate to the Discover Tab
- Scroll down until you come to "Topics you like" section
- Note the following:
-- The card is no longer vertical text
-- Chips are shown
-- The chip colors follows the pattern: Green, Blue, Yellow and Orange
-- You can still scroll down & up
-- You can still page left & right

PR submission checklist:

- [X] I have considered adding unit tests where possible.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

zwarm committed:
Add layout margins

zwarm committed:
Update layout to use chip and add style

zwarm committed:
Remove divider and add ChipStyle

zwarm committed:
Change layoutManager to horizontal. Add onItemtouchListener

zwarm committed:
Update item viewHolder and build list to support chip styles

zwarm committed:
Add and modify tests

zwarm committed:
Fix ktlint - remove unused style

ashiagr reviewed (commented):
Thanks @zwarm! Interest chips look great 👍.

Two findings: 

1. On clicking an interest, I wasn't able to go to details screen (Tested on Pixel 5).
2. Horizontal scrolling interests was sometimes detected as a swipe gesture. Can we increase interests section touch height?
 
https://user-images.githubusercontent.com/1405144/109614096-28fdeb00-7b58-11eb-929e-d507ccebc585.mp4

zwarm committed:
Increase target size of chip items

zwarm committed:
Move addOnItemTouchListener to bind from init

zwarm committed:
Remove buildIsStartingChip

zwarm commented:
@ashiagr - Thanks so much for the review.

> 1. On clicking an interest, I wasn't able to go to details screen (Tested on Pixel 5).

This one is a bit perplexing and nice catch, I had been testing on my emulator running 10. I had to move the setting of the `onItemtouchListener` out of init and into bind. I'm still Googling for why that is the case. I tested on my Samsung S20 running 11 and a few emulators with a bunch of different versions. Would love your thoughts on this one. Thanks.

> 2. Horizontal scrolling interests was sometimes detected as a swipe gesture. Can we increase interests section touch height?

I removed the recycler top/bottom margins and added them as padding to the item, thus increasing the grab area. Let me know what you think, based on designs this matches the spacing requests by Chris.

I also removed the `fun buildIsStartingChip`, as it was no longer being used.

zwarm commented:
FYi @mbshakti :)

ashiagr commented:
Thanks @zwarm!

> This one is a bit perplexing and nice catch, I had been testing on my emulator running 10. I had to move the setting of the onItemtouchListener out of init and into bind. I'm still Googling for why that is the case. I tested on my Samsung S20 running 11 and a few emulators with a bunch of different versions. Would love your thoughts on this one. Thanks.

I retested after new changes and still can't open details screen. I always get `MotionEvent.ACTION_MOVE ` event when I touch an interest chip on Pixel 5, and `MotionEvent.ACTION_DOWN` on an emulator. Seems something related to touch sensitivity. Can we add some threshold distance to detect whether it is an `ACTION_MOVE` or `ACTION_DOWN`  event?

> I removed the recycler top/bottom margins and added them as padding to the item, thus increasing the grab area. Let me know what you think, based on designs this matches the spacing requests by Chris.

Looks better now 👍.

EDIT: Looks like fling gesture still triggers swipe on viewpager even though I was within the bounds on swiping/ dragging.
Not a blocker, but we might want to look into it.

https://user-images.githubusercontent.com/1405144/109763258-d7b53080-7c17-11eb-8f59-58543ed68aea.mp4

zwarm commented:
> I retested after new changes and still can't open details screen. I always get `MotionEvent.ACTION_MOVE ` event when I touch an interest chip on Pixel 5, and `MotionEvent.ACTION_DOWN` on an emulator. Seems something related to touch sensitivity. Can we add some threshold distance to detect whether it is an `ACTION_MOVE` or `ACTION_DOWN` event?

Ugh, I can't recreate this with my physical devices (3XL & Samsung S20) or on an emulator - I am able to click through. I'll spend the day on setting up that threshold, but I'll need your help with testing.


> EDIT: Looks like fling gesture still triggers swipe on viewpager even though I was within the bounds on swiping/ dragging.
I'll look into the fling gestures - I might have to completely write a custom listener. 

Thanks.

zwarm committed:
Change left-right padding on chip

zwarm committed:
Use chip onClickListener instead of container

zwarm committed:
Add color resources for reader topic chips

zwarm committed:
Add color resources for reader topic chips

zwarm committed:
Support updated chip color resources

zwarm committed:
make start-end margin smaller

zwarm committed:
Change ReaderInterestChipStyleColor.BLUE to PURPLE

ashiagr commented:
Thanks @zwarm! 

I re-tested and Pixel 5 interest chip issue is resolved now. As discussed, we can address ViewPager + ScrollView scrolling conflict in another PR.

Feel free to merge this PR after design review.

mbshakti commented:
hey @zwarm this is the latest ui for the chips, right? just wanted to confirm before approving the pr!

<img width="802" alt="Screen Shot 2021-03-05 at 3 33 00 PM" src="https://user-images.githubusercontent.com/6350997/110170353-1853ac00-7dc8-11eb-87b1-7668c9d9309d.png">
